File: solutions/day7.py
with open("./input.txt", "r") as txt:
  raw: str = txt.read().rstrip()
  txt.close()

data = list(map(list, zip(*[_.split(': ') for _ in raw.split('\n')])))
values, equations = list(map(int, data[0])), [list(map(int, _.split(' '))) for _ in data[1]]

# Part 1
resultsPart1 = 0
for i in range(len(values)):
    base10 = 2 ** (len(equations[i]) - 1)
    operators = list(map(int, f'{base10:b}'[1:]))
    while True:
        result = equations[i][0]
        for j in range(len(operators)):
            if operators[j] == 0:
                result += equations[i][j + 1]
            else:
                result *= equations[i][j + 1]
        if result == values[i]:
            resultsPart1 += values[i]
            # print(operators)
            break
        # print(result, operators)
        base10 += 1
        if len(list(map(int, f'{base10:b}'[1:]))) > len(operators):
            break
        operators = list(map(int, f'{base10:b}'[1:]))

print("Part 1:", resultsPart1)

# Part 2
# referencing https://stackoverflow.com/questions/34559663/convert-decimal-to-ternarybase3-in-python
def ternary(n: int) -> str:
    if n == 0:
        return '0'
    nums = []
    while n:
        n, r = divmod(n, 3)
        nums.append(str(r))
    return ''.join(reversed(nums))

resultsPart2 = 0
for i in range(len(values)):
    base10 = 3 ** (len(equations[i]) - 1)
    operators = list(map(int, ternary(base10)[1:]))
    while True:
        result = equations[i][0]
        for j in range(len(operators)):
            if operators[j] == 0:
                result += equations[i][j + 1]
            elif operators[j] == 1:
                result *= equations[i][j + 1]
            else:
                result = int(str(result) + str(equations[i][j + 1]))
        if result == values[i]:
            resultsPart2 += values[i]
            # print(operators)
            break
        # print(result, operators)
        base10 += 1
        if len(list(map(int, ternary(base10)[1:]))) > len(operators):
            break
        operators = list(map(int, ternary(base10)[1:]))

print("Part 2:", resultsPart2)

File: solutions/day7.ts
import fs from "node:fs";

const raw = fs.readFileSync("./input.txt", "utf-8");
const rawData = raw.split("\n").map((line) => line.split(": "));
const data = rawData[0].map((_, colIndex) =>
  rawData.map((row) => row[colIndex])
);
const values = data[0].map(Number);
const equations = data[1].map((equation) => equation.split(" ").map(Number));

// Part 1
const resultsPart1 = values.reduce((acc, value, i) => {
  while (true) {
    const tryOperators = (base10: number): boolean => {
      const operators = base10.toString(2).split("").slice(1).map(Number);
      const result = operators.reduce((accResult, operator, j) => {
        if (operator === 0) {
          return accResult + equations[i][j + 1];
        } else {
          return accResult * equations[i][j + 1];
        }
      }, equations[i][0]);
      if (result === values[i]) {
        return true;
      }
      if (
        (base10 + 1).toString(2).split("").slice(1).length > operators.length
      ) {
        return false;
      }
      return tryOperators(base10 + 1);
    };
    if (tryOperators(2 ** (equations[i].length - 1))) {
      return acc + value;
    } else {
      return acc;
    }
  }
}, 0);
console.log("Part 1:", resultsPart1);

// Part 2
const resultsPart2 = values.reduce((acc, value, i) => {
  let base10 = 3 ** (equations[i].length - 1);
  while (true) {
    const operators = base10.toString(3).split("").slice(1).map(Number);
    const result = operators.reduce((accResult, operator, j) => {
      if (operator === 0) {
        return accResult + equations[i][j + 1];
      } else if (operator === 1) {
        return accResult * equations[i][j + 1];
      } else {
        return Number(accResult.toString() + equations[i][j + 1].toString());
      }
    }, equations[i][0]);
    if (result === values[i]) {
      return acc + value;
    }
    base10 += 1;
    if (base10.toString(3).split("").slice(1).length > operators.length) {
      return acc;
    }
  }
}, 0);
console.log("Part 2:", resultsPart2);
